server/index.js

```diff
@@ -53,6 +59,170 @@ app.get('/api/products/:productId', (req, res, next) => {
     .catch(err => next(err));
 });
 
+app.post('/api/cartItems', (req, res, next) => {
+  const { productId, customizations, quantity } = req.body;
+  const qtyNum = parseInt(quantity, 10);
+  const prodNum = parseInt(productId, 10);
+  if (!productId) {
+    throw new ClientError(400, 'productId, customizations, and quantity are required.');
+  } else if (!Number.isInteger(qtyNum) || !Number.isInteger(prodNum)) {
+    throw new ClientError(400, 'quantity and productId need to be positive integers.');
+  }
+  const cartToken = req.headers['x-access-token'];
+  Promise
+    .resolve()
+    .then(() => {
+      if (cartToken !== 'null') {
+        const payload = jwt.verify(cartToken, process.env.TOKEN_SECRET);
+        const cartIdAndToken = { cartId: payload.cartId, token: cartToken };
+        return cartIdAndToken;
+      } else {
+        const sql = `
+    insert into "carts"
+         values (default)
+      returning *
+      `;
+        db.query(sql)
+          .then(result => {
+            const payload = { cartId: result.rows[0].cartId };
+            const token = jwt.sign(payload, process.env.TOKEN_SECRET);
+            const cartIdAndToken = { cartId: payload.cartId, token: token };
+            return cartIdAndToken;
+          });
+      }
+    })
+    .then(cartIdAndToken => {
+      const { cartId, token } = cartIdAndToken;
+      const sqlCart = `
+      select "productId",
+             "customizations",
+             "cartItemsId",
+             "quantity"
+        from "cartItems"
+       where "productId" = $1;
+       `;
+      const params = [productId];
+      return db.query(sqlCart, params)
+        .then(result => {
+          const newItem = Object.values(customizations);
+          const numQuantity = JSON.parse(quantity);
+          for (let i = 0; i < result.rows.length; i++) {
+            const cartItemId = result.rows[i].cartItemsId;
+            const cartQuantity = result.rows[i].quantity;
+            const cartItem = Object.values(result.rows[i].customizations);
+            if (equal(cartItem, newItem)) {
+              const newQuantity = numQuantity + cartQuantity;
+              const sqlItem = `
+              update "cartItems"
+                 set "quantity" = $1
+               where "cartItemsId" = $2
+               returning *;
+              `;
+              const params = [newQuantity, cartItemId];
+              return db.query(sqlItem, params)
+                .then(res => {
+                  const cartItem = res.rows[0];
+                  return ({ cartItem, token });
+                });
```

Comment:
For this i think we can simplify things (remove the loop and custom comparison logic) with two moves:
1. change the data type in postgres for `customizations` to `jsonb`. this will give you access to a [handy comparison operator for comparing JSON values (`@>` and `<@`)](https://www.postgresql.org/docs/10/functions-json.html). Be prepared to cast one of your params to `jsonb` in the query (as in `where "customizations" @> $2::jsonb and "customizations" <@ $2::jsonb`
2. Try updating the quantity _first_ where the `cartId`, `productId`, and `customizations` are equal and use a `returning` clause to get the updated cart item. If none was updated then you know for sure that you need to insert a row.